src/storage.rs

```diff
@@ -1,296 +1,494 @@
-    alive: Arc<AtomicBool>,
+pub struct ReaderId<T: 'static> {
+    id: usize,
+    marker: PhantomData<&'static [T]>,
+    // stupid way to make this `Sync`,
+    // never actually locked
+    drop_notifier: Mutex<Sender<usize>>,
 }
 
-/// This static value helps assign unique ids to every storage which are then propagated to
-/// registered reader IDs, preventing reader IDs from being used with the wrong storage.
-/// It's very important to prevent this because otherwise the unsafe code used for reading
-/// could cause memory corruption.
-static RING_BUFFER_ID: AtomicUsize = ATOMIC_USIZE_INIT;
+impl<T: 'static> Drop for ReaderId<T> {
+    fn drop(&mut self) {
+        let _ = self.drop_notifier.get_mut().send(self.id);
+    }
+}
 
-/// Ring buffer, holding data of type `T`
 #[derive(Debug)]
-pub struct RingBufferStorage<T> {
-    pub(crate) data: Vec<T>,
-    buffer_id: usize,
-    reader_internal: Vec<UnsafeCell<InternalReaderId>>,
-    write_index: usize,
-    written: usize,
-    reset_written: usize,
-    /// Used for caching the current longest reader so we only have to check that reader for
-    /// determining if growth is necessary.
-    current_longest_reader: usize,
+struct ReaderMeta {
+    /// Free ids
+    free: Vec<usize>,
+    readers: Vec<Reader>,
+}
+
+impl ReaderMeta {
+    fn alloc(&mut self, last_index: usize, generation: usize) -> usize {
+        match self.free.pop() {
+            Some(id) => {
+                self.readers[id].last_index = last_index;
+                self.readers[id].generation = generation;
+
+                id
+            }
+            None => {
+                let id = self.readers.len();
+                self.readers.push(Reader {
+                    generation,
+                    last_index,
+                });
+
+                id
+            }
+        }
+    }
+
+    fn remove(&mut self, id: usize) {
+        self.readers[id].set_inactive();
+        self.free.push(id);
+    }
+
+    fn nearest_index(&self, last: CircularIndex, current_gen: usize) -> Option<&Reader> {
+        self.readers
+            .iter()
+            .filter(|reader| reader.active())
+            .min_by_key(|reader| reader.distance_from(last, current_gen))
+    }
+
+    fn shift(&mut self, last_index: usize, current_gen: usize, grow_by: usize) {
+        for reader in &mut self.readers {
+            let reader = reader as &mut Reader;
+            if !reader.active() {
+                continue;
+            }
+
+            if reader.needs_shift(last_index, current_gen) {
+                reader.last_index += grow_by;
+            }
+        }
+    }
 }
 
-unsafe impl<T> Sync for RingBufferStorage<T>
-where
-    T: Sync,
-{
+/// Ring buffer, holding data of type `T`.
+#[derive(Debug)]
+pub struct RingBuffer<T> {
+    available: usize,
+    last_index: CircularIndex,
+    data: Data<T>,
+    // stupid way to make this `Sync`,
+    // never actually locked
+    free_rx: Mutex<Receiver<usize>>,
```

Comment:
Comment ^

Reply:
yeah I think it's a useful hint for somebody who reads the code? It's only as long as we don't have a better, clean solution.